keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,960 @@
+    - --kubeconfig=/etc/kubernetes/scheduler.conf
+    - --leader-elect=true
+    - --port=0
+    image: k8s.gcr.io/kube-scheduler:v1.21.0-beta.0.368_9850bf06b571d5-dirty
+    imagePullPolicy: IfNotPresent
+    livenessProbe:
+      failureThreshold: 8
+      httpGet:
+        host: 127.0.0.1
+        path: /healthz
+        port: 10259
+        scheme: HTTPS
+      initialDelaySeconds: 10
+      periodSeconds: 10
+      timeoutSeconds: 15
+    name: kube-scheduler
+    resources:
+      requests:
+        cpu: 100m
+    securityContext:
+      allowPrivilegeEscalation: false
+      capabilities:
+        drop:
+        - ALL
+      runAsGroup: 2000
+      runAsUser: 2000
+    startupProbe:
+      failureThreshold: 24
+      httpGet:
+        host: 127.0.0.1
+        path: /healthz
+        port: 10259
+        scheme: HTTPS
+      initialDelaySeconds: 10
+      periodSeconds: 10
+      timeoutSeconds: 15
+    volumeMounts:
+    - mountPath: /etc/kubernetes/scheduler.conf
+      name: kubeconfig
+      readOnly: true
+  hostNetwork: true
+  priorityClassName: system-node-critical
+  volumes:
+  - hostPath:
+      path: /etc/kubernetes/scheduler.conf
+      type: FileOrCreate
+    name: kubeconfig
+status: {}
+```
+The initContainer `kube-scheduler-init` will run before the `kube-scheduler` container and will setup the permissions of the files that the `kube-scheduler` container needs. Since initContainers are shotlived and exit once they are done the risk from running them as root on the host is low.
+
+2. **kubeadm sets the permissions:** In this approach kubeadm would be responsible for setting the file permissions when it creates the files. It will call [os.Chown](https://golang.org/pkg/os/#Chown) to set the owner of the files. This approach is demonstrated in PR https://github.com/kubernetes/kubernetes/pull/99753.
+
+The author(s) believe that it is better for kubeadm to set the permission beacuse adding an initContainer would require pulling of debian-base image (or similar) to run the commands to change file ownership, which is something that can be easily done in go. Also it leads to the question of which initContainer would be responsible for files shared between kube-controller-manager and kube-apiserver? Since kubeadm creates these files its best for it to apply the file permissions.
+
+### Shared files
+Certain hostVolume mounts are shared between kube-apiserver and kube-controller-manager, some examples of these are:-
+- /etc/kubernetes/pki/ca.crt
+- /etc/kubernetes/pki/sa.key
+
+We propose that files shared by kube-controller-manager and kube-apiserver be readable by a particular GID and we set the `supplementalGroups` in the [PodSecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritycontext-v1beta1-policy) of the Pod to this GID for kube-apiserver and kube-controller-manager.
+
+For instance lets consider that kubeadm grant group 2100 read permissions to /etc/kubernetes/pki/ca.crt
```

Comment:
what if group 2100 is taken by something else on the host?

Reply:
If we hardcode uid/gids then there is no good way to determine this, but if we let the user set it then it is their responsibility to ensure that these are not assigned to anyone else. See the "How do we assign these uid/gids to the components?" section.